syntax = "proto3";

service EdgeService {

  rpc Echo (echoMessage) returns (echoMessage) {}

  rpc InitBench(InitBenchRequest) returns (InitBenchResponse) {}

  rpc StartTraining(InitTrainRequest) returns (InitTrainResponse) {}

  rpc StreamFile(stream UploadFile) returns (StringResponse) {}

  rpc StartValidation(InitValidationRequest) returns (InitValidationResponse) {}

}

message MetaData {
  string model_id = 1;
  string file_name = 2;
}

message UploadFile {
  oneof request {
    MetaData metadata = 1;
    bytes chunk_data = 2;
  }
}

message File{
  bytes chunk_data=1;
}

message StringResponse{
  string text=1;
}

message echoMessage {
  string text=1;
}

message InitBenchRequest {
  string model_id = 1;
  string model_class= 2;
  bytes model_config= 3;
  string dataset_id = 4;
  int32 batch_size = 5;
  float learning_rate= 6;
  optional bytes optimizer = 7;
  optional bytes loss_function = 8;
  oneof request{
    float timeout_duration_s = 9;
    int32 max_mini_batch_count = 10;
  }
}

message InitTrainRequest {
  string session_id = 1;
  string model_id = 2;
  string model_class= 3;
  bytes model_config= 4;
  string dataset_id = 5;
  bytes model_wts= 6;
  int32 batch_size = 7;
  float learning_rate= 8;
  int32 num_epochs = 9;
  int32 round_idx = 10;
  optional bytes optimizer = 11;
  optional bytes loss_function = 12;
  oneof request{
    float timeout_duration_s = 13;
    int32 max_mini_batch_count = 14;
  }
}

message InitValidationRequest{
  string session_id = 1;
  string model_id = 2;
  string model_class= 3;
  bytes model_config = 4;
  string dataset_id = 5;
  bytes model_wts = 6;
  int32 batch_size = 7;
  int32 round_idx = 8;
  optional bytes optimizer = 9;
  optional bytes loss_function = 10;
}

message InitBenchResponse {
  string model_id = 1;
  int32 num_mini_batches = 2;
  float bench_duration_s = 3;
}

message InitTrainResponse{
  string model_id = 1;
  bytes model_weights = 2;
  string client_id = 3;
  int32 round_idx = 4;
  bytes metrics = 5;
}

message InitValidationResponse{
  string model_id = 1;
  string client_id = 2;
  int32 round_idx = 3;
  bytes metrics = 4;
}